src/domains/ValuesDomain.js

```diff
@@ -248,6 +252,112 @@ export default class ValuesDomain {
+      let oldValue = ToInt32(realm, value);
+
+      // 3. Return the result of applying bitwise complement to oldValue. The result is a signed 32-bit integer.
+      return new NumberValue(realm, ~oldValue);
+    } else if (op === "!") {
+      // ECMA262 12.6.9
+      // 1. Let expr be the result of evaluating UnaryExpression.
+      // 2. Let oldValue be ToBoolean(? GetValue(expr)).
+      let oldValue = ToBoolean(realm, value);
+
+      // 3. If oldValue is true, return false.
+      if (oldValue === true) return realm.intrinsics.false;
+
+      // 4. Return true.
+      return realm.intrinsics.true;
+    } else if (op === "void") {
+      // 1. Let expr be the result of evaluating UnaryExpression.
+      // 2. Perform ? GetValue(expr).
+      // 3. Return undefined.
+      return realm.intrinsics.undefined;
+    } else if (op === "typeof") {
+      function isInstance(proto, Constructor): boolean {
+        return proto instanceof Constructor || proto === Constructor.prototype;
+      }
+      // ECMA262 12.6.5
+      // 1. Let val be the result of evaluating UnaryExpression.
+      // 2. If Type(val) is Reference, then
+      // 3. Let val be ? GetValue(val).
+      let val = value;
+      // 4. Return a String according to Table 35.
+      let proto = val.getType().prototype;
+      if (isInstance(proto, UndefinedValue)) {
+        return new StringValue(realm, "undefined");
+      } else if (isInstance(proto, NullValue)) {
+        return new StringValue(realm, "object");
+      } else if (isInstance(proto, StringValue)) {
+        return new StringValue(realm, "string");
+      } else if (isInstance(proto, BooleanValue)) {
+        return new StringValue(realm, "boolean");
+      } else if (isInstance(proto, NumberValue)) {
+        return new StringValue(realm, "number");
+      } else if (isInstance(proto, SymbolValue)) {
+        return new StringValue(realm, "symbol");
+      } else if (isInstance(proto, ObjectValue)) {
+        if (IsCallable(realm, val)) {
+          return new StringValue(realm, "function");
+        }
+        return new StringValue(realm, "object");
+      } else {
+        invariant(false);
+      }
+    } else {
+      invariant(op === "delete");
+      // ECMA262 12.5.3.2
+      // 1. Let ref be the result of evaluating UnaryExpression.
+      // 2. ReturnIfAbrupt(ref).
+      // 3. If Type(ref) is not Reference, return true.
+      return realm.intrinsics.true;
+    }
+  }
+
+  static unaryOp(realm: Realm, op: BabelUnaryOperator, operandValues: ValuesDomain): ValuesDomain {
+    let operandElements = operandValues._elements;
+    if (operandElements === undefined) return ValuesDomain.topVal;
+    let resultSet = new Set();
+    let savedHandler = realm.errorHandler;
+    try {
+      realm.errorHandler = () => {
+        throw new FatalError();
+      };
+      for (let operandElem of operandElements) {
+        let result = ValuesDomain.computeUnary(realm, op, operandElem);
```

Comment:
It seems to me that this can run arbitrary (state-mutating) code (ToNumber calls ToPrimitive which can call anything on an object), and there's nothing guarding that here, or even a comment about it.

Reply:
Good point. I've added some guards and a comment.